source/fuzz/fuzzer_pass_changing_memory_scope.cpp

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2021 Mostafa Ashraf
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/fuzzer_pass_changing_memory_scope.h"
+
+#include "source/fuzz/fuzzer_context.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "source/fuzz/protobufs/spirvfuzz_protobufs.h"
+#include "source/fuzz/transformation_changing_memory_scope.h"
+
+namespace spvtools {
+namespace fuzz {
+
+FuzzerPassChangingMemoryScope::FuzzerPassChangingMemoryScope(
+    opt::IRContext* ir_context, TransformationContext* transformation_context,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations,
+    bool ignore_inapplicable_transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations, ignore_inapplicable_transformations) {}
+
+void FuzzerPassChangingMemoryScope::Apply() {
+  ForEachInstructionWithInstructionDescriptor(
+      [this](opt::Function* /* unused */, opt::BasicBlock* /* unused */,
+             opt::BasicBlock::iterator inst_it,
+             const protobufs::InstructionDescriptor& instruction_descriptor) {
+        if (!GetFuzzerContext()->ChoosePercentage(
+                GetFuzzerContext()->GetChanceOfChangingMemoryScope())) {
+          return;
+        }
+
+        // Instruction must be atomic instruction only.
+        if (!TransformationChangingMemoryScope::IsAtomicInstruction(
```

Comment:
As discussed in our meeting, we should include barrier instructions. Perhaps rename `IsAtomicInstruction` to `HasMemorySemanticsOperand`.

Reply:
Done. I believe here you meant `HasMemoryScopesOperand`.

Reply:
Yes, oops. I think `HasMemoryScopeOperand` would be good (remove "s").

Reply:
Done.